streamer/src/quic.rs

```diff
@@ -0,0 +1,188 @@
+    let priv_key = cert.serialize_private_key_der();
+    let priv_key = rustls::PrivateKey(priv_key);
+    let cert_chain = vec![rustls::Certificate(cert_der)];
+    Ok((cert_chain, priv_key))
+}
+
+fn new_cert_params(identity_keypair: &Keypair, san: IpAddr) -> CertificateParams {
+    // TODO(terorie): Is it safe to sign the TLS cert with the identity private key?
+
+    // Unfortunately, rcgen does not accept a "raw" Ed25519 key.
+    // We have to convert it to DER and pass it to the library.
+
+    // Convert private key into PKCS#8 v1 object.
+    // RFC 8410, Section 7: Private Key Format
+    // https://datatracker.ietf.org/doc/html/rfc8410#section-7
+    let mut private_key = Vec::<u8>::with_capacity(34);
+    private_key.extend_from_slice(&[0x04, 0x20]); // ASN.1 OCTET STRING
+    private_key.extend_from_slice(identity_keypair.secret().as_bytes());
+    let key_pkcs8 = pkcs8::PrivateKeyInfo {
+        algorithm: AlgorithmIdentifier {
+            oid: ObjectIdentifier::from_arcs(&[1, 3, 101, 112]).unwrap(),
+            parameters: None,
+        },
+        private_key: &private_key,
+        public_key: None,
+    };
+    let key_pkcs8_der = key_pkcs8
+        .to_der()
+        .expect("Failed to convert keypair to DER")
+        .to_der();
+
+    // Parse private key into rcgen::KeyPair struct.
+    let keypair =
+        rcgen::KeyPair::from_der(&key_pkcs8_der).expect("Failed to parse keypair from DER");
+
+    let mut cert_params = CertificateParams::default();
+    cert_params.subject_alt_names = vec![SanType::IpAddress(san)];
+    cert_params.alg = &rcgen::PKCS_ED25519;
+    cert_params.key_pair = Some(keypair);
+    cert_params.distinguished_name = DistinguishedName::new();
+    cert_params
+        .distinguished_name
+        .push(DnType::CommonName, "Solana node");
+    cert_params
+}
+
+fn rt() -> Runtime {
+    Builder::new_current_thread().enable_all().build().unwrap()
+}
+
+pub fn spawn_server(
+    sock: UdpSocket,
+    keypair: &Keypair,
+    gossip_host: IpAddr,
+    packet_sender: Sender<PacketBatch>,
+    exit: Arc<AtomicBool>,
+) -> Result<thread::JoinHandle<()>, Box<dyn Error>> {
+    let (config, _cert) = configure_server(keypair, gossip_host)?;
+    let handle = thread::spawn(move || {
+        let runtime = rt();
+        let (_, mut incoming) = {
+            let _guard = runtime.enter();
+            Endpoint::new(EndpointConfig::default(), Some(config), sock).unwrap()
+        };
+        let handle = runtime.spawn(async move {
+            while !exit.load(Ordering::Relaxed) {
+                let timeout_connection =
+                    timeout(Duration::from_millis(1000), incoming.next()).await;
+
+                if let Ok(connection) = timeout_connection {
+                    let quinn::NewConnection {
+                        connection,
+                        mut uni_streams,
+                        ..
+                    } = connection.unwrap().await.expect("connect");
```

Comment:
For type clarity, I think this would be better written as:

```
let connecting: quinn::Connecting = connection.unwrap();
                    let connecting_result: Result<quinn::NewConnection, quinn::ConnectionError> = connecting.await;
                    let quinn::NewConnection {
                        connection,
                        mut uni_streams,
                        ..
                    } = connecting_result.expect("connect");
```

Reply:
Written like so, it's clearer that the `connecting_result.expect("connect");` is not safe because of the possible `ConnectionError`'s that can be returned: https://docs.rs/quinn/latest/quinn/enum.ConnectionError.html

Reply:
the first `let connecting: quinn::Connecting = connection.unwrap();` should also probably be handled